---
title: "PBMC Single-Cell RNA Sequencing Data Analysis Pipeline"
#author: "Wenke Sun did this file"
date: "2025-02-13"
output:
  html_document:
    toc: true
    toc_depth: 3
    toc_float: true
---

# Project Name: PBMC Single-Cell RNA Sequencing Data Analysis Pipeline

## Table of Contents
1. [Part 1: Data Load](#part-1-data-load)
   - [Project Description](#project-description)
   - [Code Structure](#code-structure)
   - [Usage Instructions](#usage-instructions)
   - [Code Example](#code-example)
2. [Part 2: Wavelet Analysis](#part-2-wavelet-analysis)
   - [Overview](#overview)
   - [Function: `wavelet_transform`](#function-wavelet_transform)
3. [Part 3: Single-Cell RNA-Seq Analysis Pipeline](#part-3-single-cell-rna-seq-analysis-pipeline)
   - [Overview](#overview-1)
   - [Functions](#functions)
4. [Part 4:Clustering Methods](#clustering-methods)
   - [HGC (Hierarchical Graph Clustering)](#part-4-hgc-hierarchical-graph-clustering)
   - [SC3 (Single-Cell Consensus Clustering)](#part-5-sc3-single-cell-consensus-clustering)
5. [Part 5:Evaluation Metrics](#evaluation-metrics)
   - [ARI](#ari-adjusted-rand-index)
   - [NMI](#nmi-normalized-mutual-information)
6. [Dependencies](#dependencies)
7. [License](#license)
8. [Contributions](#contributions)
9. [Contact](#contact)

---

## Part 1: Data Load {#part-1-data-load}

### Project Description {#project-description}
This project provides a comprehensive pipeline for analyzing single-cell RNA sequencing data, covering data loading, quality control, normalization, and preliminary visualization. The code utilizes the Seurat package, a widely used R package for single-cell RNA sequencing data analysis.

### Code Structure {#code-structure}
- **Data Loading**: Load the PBMC (Peripheral Blood Mononuclear Cells) dataset from a specified directory.
- **Quality Control**: Filter cells by calculating the percentage of mitochondrial genes and RNA feature counts.
- **Data Normalization**: Normalize the data using the LogNormalize method.
- **Visualization**: Generate violin plots and feature scatter plots to visually assess data quality.

### Usage Instructions {#usage-instructions}
1. Ensure the Seurat package is installed. If not, install it using the following command:
   ```R
   install.packages("Seurat")
   ```
2. Place the PBMC dataset in the `pbmc_1k` folder within the project directory.
3. Run the provided R script. The code will automatically load the data, perform quality control and normalization, and generate corresponding plots.

### Code Example {#code-example}
```{r, echo=FALSE, results = 'hide'}
#packages
library(dplyr)
library(Seurat)
library(patchwork)
library(wavelets)
library(Matrix)
library(scCATCH)
library(hdf5r)
library(readr)
library(SAVER)
library(parallel)
library(tidyverse)
library(HGC)
library(SingleCellExperiment)
library(SC3)
library(scater)
library(mclust)
library(randnet)
#wavelet functions
# Define the wavelet transformation function
#Arguments 
#wavelet_type "2-band","3-band","4-band"
#daubechies_order "d4","d8","d12" or "d16"
#regular True if regular = 4, False regular = 2
wavelet_transform <- function(input_matrix, wavelet_type , daubechies_order = "d4", regular = FALSE) {
  
  # 2-band wavelet transform
  if (wavelet_type == "2-band") {
    # Check if the Daubechies order is valid
    if (!daubechies_order %in% c("d4","d8","d12","d16")) {
      stop("For 2-band wavelet transform, Daubechies order must be 4, 8, 12, or 16.")
    }
    
    # Perform 2-band wavelet transform
    
    result <- two_band(input_matrix,daubechies_order)
  }
  
  # 3-band wavelet transform
  else if (wavelet_type == "3-band") {
    # Perform 3-band wavelet transform
    result <- three_band(input_matrix)
  }
  
  # 4-band wavelet transform
  else if (wavelet_type == "4-band") {
    if (regular) {
      # Use 4-regular wavelet
      result <- four_band4(input_matrix)
    } else {
      # Use default 4-band wavelet
      result <- four_band(input_matrix)
    }
  }
  
  # If the wavelet type is invalid
  else {
    stop("Unsupported wavelet type. Please choose '2-band', '3-band', or '4-band'.")
  }
  
  # Return the wavelet transform result
  return(result)
}
#2-band wavelet
two_band <- function(X,daubechies_order){
  n=dim(X)[1]
  m=dim(X)[2]
  VV=matrix(NA,n,m)
  WW=matrix(NA,n,m)
  #d8 d12
  for (i in 1:m){
    p<-X[,i]
    B<-dwt(p,daubechies_order,n.level=1)
    
    B1<-B
    B2<-B
    B1@W$W1<-matrix(0,n/2,1)
    B2@V$V1<-matrix(0,n/2,1)
    VV[,i]<-idwt(B1)
    WW[,i]<-idwt(B2)
  }
  
  result= list("low" = VV, "high" = WW)
  return (result)
}
#3-band wavelet
three_band <- function(X){
  n=dim(X)[1]
  m=dim(X)[2]
  h0=c(0.33838609728386,0.53083618701374,0.72328627674361,0.23896417190576,0.04651408217589,-0.14593600755399)
  h1=c(-0.11737701613483,0.54433105395181,-0.01870574735313,-0.69911956479289,-0.13608276348796,0.42695403781698)
  h2=c(0.40363686892892,-0.62853936105471,0.46060475252131,-0.40363686892892,-0.07856742013185,0.24650202866523)
  #n=17125  ##number of genes
  #m=2359  ##number of cells
  nn=3*ceiling(n/3)
  
  vv1=matrix(0,nn,m)
  ww1=matrix(0,nn,m)
  ww2=matrix(0,nn,m)
  for (k in 1:m){
    ss=X[,k]
    if(n%%3){ss=c(ss,rep(0,3-n%%3))}
    ss=c(ss,ss[1:3])
    v1=rep(0,nn/3)
    w1=rep(0,nn/3)
    w2=rep(0,nn/3)
    j=1
    for (i in 1:(nn/3)){
      v1[i]=sum(h0*ss[j:(j+5)])
      w1[i]=sum(h1*ss[j:(j+5)])
      w2[i]=sum(h2*ss[j:(j+5)])
      j=j+3
    }
    v1=c(v1[nn/3],v1)
    w1=c(w1[nn/3],w1)
    w2=c(w2[nn/3],w2)
    
    for (i in 1:(nn/3)){
      vv1[3*i-2,k]=h0[4]*v1[i]+h0[1]*v1[i+1]
      vv1[3*i-1,k]=h0[5]*v1[i]+h0[2]*v1[i+1]
      vv1[3*i,k]=h0[6]*v1[i]+h0[3]*v1[i+1]
      ww1[3*i-2,k]=h1[4]*w1[i]+h1[1]*w1[i+1]
      ww1[3*i-1,k]=h1[5]*w1[i]+h1[2]*w1[i+1]
      ww1[3*i,k]=h1[6]*w1[i]+h1[3]*w1[i+1]
      ww2[3*i-2,k]=h2[4]*w2[i]+h2[1]*w2[i+1]
      ww2[3*i-1,k]=h2[5]*w2[i]+h2[2]*w2[i+1]
      ww2[3*i,k]=h2[6]*w2[i]+h2[3]*w2[i+1]
    }
  }
  vv1=vv1[1:n,]
  ww1=ww1[1:n,]
  ww2=ww2[1:n,]
  result= list("low" = vv1, "high1" = ww1,'high2'=ww2)
  return (result)
}

#4-band wavelet
four_band <- function(X){
  n=dim(X)[1]
  m=dim(X)[2]
  h0<-(c(-0.067371764,0.094195111,0.40580489,0.567371764,0.567371764,0.40580489,0.094195111,-0.067371764))
  h1<-(c(-0.094195111,0.067371764, 0.567371764 ,0.40580489,-0.40580489,-0.567371764,-0.067371764,0.094195111))
  h2<-(c(-0.094195111,-0.067371764,0.567371764,-0.40580489,-0.40580489,0.567371764,-0.067371764,-0.094195111))
  h3<-(c(-0.067371764,-0.094195111,0.40580489,-0.567371764,0.567371764,-0.40580489,0.094195111,0.067371764))
  
  #n=17125  ##number of genes
  #m=2359  ##number of cells
  nn=4*ceiling(n/4)
  
  vv1=matrix(0,nn,m)
  ww1=matrix(0,nn,m)
  ww2=matrix(0,nn,m)
  ww3=matrix(0,nn,m)
  for (k in 1:m){
    ss=X[,k]
    if(n%%4){ss=c(ss,rep(0,4-n%%4))}
    ss=c(ss,ss[1:4])
    v1=rep(0,nn/4)
    w1=rep(0,nn/4)
    w2=rep(0,nn/4)
    w3=rep(0,nn/4)
    j=1
    for (i in 1:(nn/4)){
      v1[i]=sum(h0*ss[j:(j+7)])
      w1[i]=sum(h1*ss[j:(j+7)])
      w2[i]=sum(h2*ss[j:(j+7)])
      w3[i]=sum(h3*ss[j:(j+7)])
      j=j+4
    }
    v1=c(v1[nn/4],v1)
    w1=c(w1[nn/4],w1)
    w2=c(w2[nn/4],w2)
    w3=c(w3[nn/4],w3)
    for (i in 1:(nn/4)){
      vv1[4*i-3,k]=h0[5]*v1[i]+h0[1]*v1[i+1]
      vv1[4*i-2,k]=h0[6]*v1[i]+h0[2]*v1[i+1]
      vv1[4*i-1,k]=h0[7]*v1[i]+h0[3]*v1[i+1]
      vv1[4*i,k]=h0[8]*v1[i]+h0[4]*v1[i+1]
      ww1[4*i-3,k]=h1[5]*w1[i]+h1[1]*w1[i+1]
      ww1[4*i-2,k]=h1[6]*w1[i]+h1[2]*w1[i+1]
      ww1[4*i-1,k]=h1[7]*w1[i]+h1[3]*w1[i+1]
      ww1[4*i,k]=h1[8]*w1[i]+h1[4]*w1[i+1]
      ww2[4*i-3,k]=h2[5]*w2[i]+h2[1]*w2[i+1]
      ww2[4*i-2,k]=h2[6]*w2[i]+h2[2]*w2[i+1]
      ww2[4*i-1,k]=h2[7]*w2[i]+h2[3]*w2[i+1]
      ww2[4*i,k]=h2[8]*w2[i]+h2[4]*w2[i+1]
      ww3[4*i-3,k]=h3[5]*w3[i]+h3[1]*w3[i+1]
      ww3[4*i-2,k]=h3[6]*w3[i]+h3[2]*w3[i+1]
      ww3[4*i-1,k]=h3[7]*w3[i]+h3[3]*w3[i+1]
      ww3[4*i,k]=h3[8]*w3[i]+h3[4]*w3[i+1]
    }
  }
  vv1=vv1[1:n,]
  ww1=ww1[1:n,]
  ww2=ww2[1:n,]
  ww3=ww3[1:n,]
  
  result= list("low" = vv1, "high1" = ww1,'high2'=ww2,'high3'=ww3)
  return (result)
}



main_function<-function(data_,ob,X,t1,t2){
  pbmc_vv<-ob
  colnames(data_)<-colnames(X)
  rownames(data_)<-rownames(X)
  vv_sprase<-Matrix(data_, sparse = TRUE)
  pbmc_vv@assays$RNA$data<-vv_sprase
  pbmc_vv <- FindVariableFeatures(pbmc_vv, selection.method = "vst", nfeatures = 5000)
  top10_vv <- head(VariableFeatures(pbmc_vv), 10)
  
  plot1 <- VariableFeaturePlot(pbmc_vv)
  plot2 <- LabelPoints(plot = plot1, points = top10_vv, repel = TRUE)
  plot1 + plot2
  
  all.genes <- rownames(pbmc_vv)
  pbmc_vv <- ScaleData(pbmc_vv, features = all.genes)
  
  pbmc_vv <- RunPCA(pbmc_vv, features = VariableFeatures(object = pbmc_vv))
  
  VizDimLoadings(pbmc_vv, dims = 1:2, reduction = "pca")
  DimPlot(pbmc_vv, reduction = "pca")
  pbmc_vv <- JackStraw(pbmc_vv, num.replicate = 100)
  pbmc_vv <- ScoreJackStraw(pbmc_vv, dims = 1:20)
  
  JackStrawPlot(pbmc_vv, dims = 1:15)
  ElbowPlot(pbmc_vv)
  
  pbmc_vv <- FindNeighbors(pbmc_vv, dims = 1:10)
  pbmc_vv <- FindClusters(pbmc_vv, resolution = 0.5)
  head(Idents(pbmc_vv), 10)
  pbmc_vv <- RunUMAP(pbmc_vv, dims = 1:10)
  
  
  pbmc_vv.markers <- FindAllMarkers(pbmc_vv, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25)
  

  
  result= list("data" = pbmc_vv, "markers" = pbmc_vv.markers)
  return(result)
}


##########fix for different disease:FLASE, Logfc=0.2

cell_identity<-function(data,tissue_name,match_CellMatch_,cancer){
  clu_markers <- findmarkergenes(object = data,species = 'Human',cluster = 'All',match_CellMatch = match_CellMatch_,
                                 cancer = cancer,
                                 tissue = tissue_name,
                                 cell_min_pct = 0.1,#0.25,
                                 logfc = 0.1,#0.25,
                                 pvalue = 0.01)#0.05
  
  clu_ann <- scCATCH(object = clu_markers$clu_markers,
                     species = 'Human',
                     cancer = cancer,
                     tissue = tissue_name)
  
  result= list("clu_markers " = clu_markers , "indenity" =clu_ann )
  return (result)
}
#############


plot_umap<-function(data,list_name){
  new.cluster.ids <- list_name
  names(new.cluster.ids) <- levels(data)
  data<- RenameIdents(data, new.cluster.ids)
  
  return (DimPlot(data, reduction = "umap", label = T, pt.size = 0.8,label.size = 2)) 
}



cell_type_name<-function(d1,d2){
  
  
  name=d1$indenity$cell_type
  d3=d2$data@meta.data$seurat_clusters
  len_=length(name)
  for(i in 1:len_){
    if (is.na(name[i])){
      name[i]='Unknown'
    }
  }
  
  a=as.numeric(unique(d3))-1
  new_name=1:max(a)
  b=as.numeric(d1$indenity$cluster)
  c=setdiff(a,b)+1
  
  if (length(c)!=0){
    for (j in c){
      new_name[j]='Unknown'
    }
    
    k=1
    for (j in (b+1)){
      new_name[j]=name[k]
      k=k+1
    }
    return (new_name)
  }
  else{
    return (name)
  }
}


plothgc=function(result){
  pbmctree=result@graphs$ClusteringTree
  pbmctree$height = log(pbmctree$height + 1)
  clus=result$seurat_clusters
  pbmclabels=data.frame(clusters=clus)
  clunum=length(levels(clus))
  HGC.PlotDendrogram(tree = pbmctree,k=clunum, plot.label=T, labels= pbmclabels)
  return(clunum)
}
```

```{r}
# Load the PBMC dataset
pbmc.data <- Read10X(data.dir = "~/Desktop/papernew/new_method/code/pbmc_1k")
pbmc <- CreateSeuratObject(counts = pbmc.data, project = "pbmc1k", min.cells = 3, min.features = 200)

# Calculate the percentage of mitochondrial genes
pbmc[["percent.mt"]] <- PercentageFeatureSet(pbmc, pattern = "^MT-")

# Visualize quality control metrics
VlnPlot(pbmc, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)

# Generate feature scatter plots
plot1 <- FeatureScatter(pbmc, feature1 = "nCount_RNA", feature2 = "percent.mt")
plot2 <- FeatureScatter(pbmc, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
plot1 + plot2

# Apply quality control filtering, min and max comes from VlnPlot
pbmc <- subset(pbmc, subset = nFeature_RNA > 200 & nFeature_RNA < 3000 & percent.mt < 5)

# Normalize the data
pbmc <- NormalizeData(pbmc, normalization.method = "LogNormalize", scale.factor = 10000)

# Prepare the data matrix for wavelet analysis
for_wavelet <- pbmc@assays$RNA$data
working_matrix <- as.matrix(for_wavelet)
```

---

## Part 2: Wavelet Analysis {#part-2-wavelet-analysis}

### Overview {#overview}
This R script defines a function `wavelet_transform` that performs wavelet transformations on an input matrix. The function supports three types of wavelet transforms: `2-band`, `3-band`, and `4-band`. 

### Function: `wavelet_transform` {#function-wavelet_transform}

#### Arguments
- `input_matrix`: A matrix on which the wavelet transformation will be performed.
- `wavelet_type`: A string specifying the type of wavelet transform. Supported values are:
  - `"2-band"`: 2-band wavelet transform.
  - `"3-band"`: 3-band wavelet transform.
  - `"4-band"`: 4-band wavelet transform.

#### Functionality
1. **2-band Wavelet Transform**:
   - Calls the `two_band` function.

2. **3-band Wavelet Transform**:
   - Calls the `three_band` function.

3. **4-band Wavelet Transform**:
   - Calls the `four_band` function.
  

4. **Error Handling**:
   - If an unsupported `wavelet_type` is provided, the function stops and returns an error message.

#### Return Value
- The function returns the result of the wavelet transformation, which is typically a transformed matrix or a list of coefficients.

#### Example Usage
```{r}
result_2band <- wavelet_transform(working_matrix, wavelet_type = "2-band")
### 2-band low frequency data first 6 rows
```

#### Notes
- Ensure that the `two_band`, `three_band`, and `four_band` functions are defined and available in your R environment.
- In our project we conduct the first level of wavelet transform, and thus the input matrix dimension should be compatible with the M-band wavelet transform. Otherwise, a few zero rows are automatically added corresponding to the number of bands. 

#### Dependencies
- This function does not rely on external packages but assumes the existence of helper functions (`two_band`, `three_band`, and `four_band`).

---

## Part 3: Single-Cell RNA-Seq Analysis Pipeline {#part-3-single-cell-rna-seq-analysis-pipeline}

### Overview {#overview-1}
This R script provides a pipeline for analyzing single-cell RNA sequencing (scRNA-seq) data. It includes functions for preprocessing, clustering, marker gene identification, and cell type annotation. The pipeline is designed to work with Seurat objects and integrates with external tools like `scCATCH` for cell type identification.

### Functions {#functions}

#### 1. `main_function`
##### Description
This function performs a comprehensive analysis of scRNA-seq data, including normalization, dimensionality reduction, clustering, and marker gene identification.

##### Arguments
- `data_`: A matrix containing the raw gene expression data.
- `ob`: A Seurat object to store the processed data.
- `X`: A reference matrix used to set the row and column names of `data_`.
- `t1`, `t2`: Optional parameters for naming output files (currently commented out).

##### Workflow
1. **Data Preprocessing**:
   - Converts the input matrix into a sparse matrix format.
   - Updates the Seurat object with the processed data.
   - Identifies highly variable genes.

2. **Visualization**:
   - Plots the top 10 variable genes.
   - Scales the data and performs PCA.

3. **Dimensionality Reduction**:
   - Visualizes PCA loadings and performs JackStraw analysis to determine significant principal components.
   - Computes UMAP for visualization.

4. **Clustering**:
   - Identifies clusters using the `FindNeighbors` and `FindClusters` functions.
   - Finds marker genes for each cluster.

5. **Output**:
   - Returns a list containing the processed Seurat object (`data`) and marker genes (`markers`).

##### Example Usage
```{r}

two_band_low<-main_function(result_2band$low,pbmc,working_matrix,'2band','low')

```

#### 2. `cell_identity`
##### Description
This function annotates cell types using the `scCATCH` package based on marker genes identified from the scRNA-seq data.

##### Arguments
- `data`: A Seurat object containing the processed scRNA-seq data.
- `tissue_name`: The tissue type of the sample (e.g., "blood", "lung").
- `match_CellMatch_`: A logical value indicating whether to match cell types using the `CellMatch` database.
- `cancer`: A logical value indicating whether the sample is from a cancer tissue.

##### Workflow
1. Identifies marker genes for each cluster using `findmarkergenes`.
2. Annotates cell types using `scCATCH`.

##### Output
- Returns a list containing:
  - `clu_markers`: Marker genes for each cluster.
  - `identity`: Annotated cell types.

##### Example Usage
```{r}
tissu_name<- 'Blood'
match_CellMatch_name<-FALSE
cancer_name<-NULL
two_band_low_cell_identity<-cell_identity(two_band_low$data,tissu_name,match_CellMatch_name, cancer_name)
```



#### 3. `cell_type_name`
##### Description
This function generates a vector of cell type names for each cluster, handling cases where cell types are unknown.

##### Arguments
- `d1`: The output of the `cell_identity` function.
- `d2`: The output of the `main_function` function.

##### Workflow
1. Extracts annotated cell types from `d1`.
2. Assigns "Unknown" to clusters without annotated cell types.
3. Returns a vector of cell type names corresponding to each cluster.

##### Output
- Returns a vector of cell type names.

##### Example Usage
```{r}
two_band_low_cell_list_name<-cell_type_name(two_band_low_cell_identity,two_band_low)

```
#### 4. `plot_umap`
##### Description
This function generates a UMAP plot with cluster labels based on annotated cell types.

##### Arguments
- `data`: A Seurat object containing the processed scRNA-seq data.
- `list_name`: A list of cell type names corresponding to each cluster.

##### Workflow
1. Renames cluster identities in the Seurat object using the provided cell type names.
2. Generates a labeled UMAP plot.

##### Output
- Returns a UMAP plot with cluster labels.

##### Example Usage
```{r}
plot_umap(two_band_low$data,two_band_low_cell_list_name)
```
---

```
# Single-Cell Clustering Workflow: HGC & SC3

This repository provides code for clustering single-cell RNA-seq data using two methods: **Hierarchical Graph-based Clustering (HGC)** and **SC3**. The workflow also includes cluster validation metrics (Adjusted Rand Index and Normalized Mutual Information).
```

## Part 4: Clustering Methods {#clustering-methods}
### Overview  {#overview}
This repository contains code for applying Hierarchical Graph-based Clustering (HGC) and SC3 clustering to single-cell RNA-seq data. The code is designed to work with Seurat objects and SingleCellExperiment objects, and it includes functions for visualizing clustering results and calculating clustering accuracy metrics such as Adjusted Rand Index (ARI) and Normalized Mutual Information (NMI).
### HGC (Hierarchical Graph Clustering) {#part-4-hgc-hierarchical-graph-clustering}
In this section, we apply the Hierarchical Graph Clustering (HGC) method to the dataset using the `FindClusteringTree` function and visualize the results using the `plothgc` function. The `FindClusteringTree` function builds a clustering tree using a similarity-based graph, and the dendrogram plot is generated to show the clustering results.
### Code Structure {#code-structure}

```{r}
# Apply HGC clustering to the dataset
pbmcvv <- FindClusteringTree(two_band_low$data, graph.type = "SNN")
plothgc(pbmcvv)


# Input: Seurat Object with HGC clustering
# Output: An HGC plot and the number of clusters
```

In the function above:
- We first generate the clustering tree using `FindClusteringTree` with an SNN (Shared Nearest Neighbor) graph type.
- We then plot the dendrogram with the `HGC.PlotDendrogram` function, which shows the hierarchical structure of clusters, and the cluster labels are shown on the plot.
- The function returns the number of clusters found by the HGC method.


### SC3 (Single-Cell Consensus Clustering){#sc3-single-cell-consensus-clustering)}
In this section, we apply the SC3 method (Single-Cell Consensus Clustering) to the dataset using the `sc3` function from the `SC3` package. SC3 performs consensus clustering on single-cell data to identify the optimal number of clusters by combining multiple clusterings into a final consensus. We then plot the consensus matrix to visualize the clustering stability.
### Code Structure {#code-structure}
```{r}
# Convert Seurat object to SingleCellExperiment object
sce = as.SingleCellExperiment(pbmc)
rowData(sce)$feature_symbol <- rownames(sce)  # Set feature names in the row data
sce@assays@data$counts = as.matrix(sce@assays@data$counts)  # Use counts data
sce@assays@data$logcounts = as.matrix(result_2band$low)  # Use log-transformed data

# Apply SC3 clustering with k clusters
k = 3  # Can be adjusted depending on the desired number of clusters
sce <- sc3(sce, ks = k, biology = TRUE)

# Plot the consensus matrix
sc3_plot_consensus(sce, k = 3)
```
In this section:
- We convert the Seurat object to a `SingleCellExperiment` object (`sce`) and assign the appropriate data.
- The `sc3` function is used to perform consensus clustering with `k` clusters, where `k` can be adjusted based on the data and the desired number of clusters.
- Finally, we visualize the consensus matrix using `sc3_plot_consensus`, which provides insight into how stable the clusters are.

### Part 5: Evaluation Metrics {#evaluation-metrics}
#### Calculating ARI and NMI (Adjusted Rand Index & Normalized Mutual Information)

To evaluate the clustering results, we calculate the **Adjusted Rand Index (ARI)** and **Normalized Mutual Information (NMI)** between the original clustering and the clustering results from the HGC and SC3 methods.
### Code Structure {#code-structure}
```{r}
orginal<-main_function(working_matrix,pbmc,working_matrix,"orginial","na")
adjustedRandIndex(two_band_low$data$seurat_clusters,orginal$data$seurat_clusters)
NMI(orginal$data$seurat_clusters,two_band_low$data$seurat_clusters)
```

## Session Information {#session-information}
Install the required packages using:
```{r}
sessionInfo()
```

---

## License {#license}
This software is distributed under the terms of the GNU General Public License, either Version 2, June 1991 or Version 3, June 2007. For more information, see the LICENSE file.

---

## Contributions {#contributions}
Feel free to raise issues or submit pull requests to improve this project.

---

## Contact {#contact}
For any questions, please contact the project maintainer.

---




